backtester/report/chart.go

```diff
@@ -0,0 +1,181 @@
+			}
+			plots = append(plots, LinePlot{
+				Value:     items[i].ReportItem.Snapshots[j].Available.InexactFloat64(),
+				UnixMilli: items[i].ReportItem.Snapshots[j].Time.UTC().UnixMilli(),
+			})
+		}
+		response.Data = append(response.Data, ChartLine{
+			Name:      fmt.Sprintf("%v %v %v holdings", items[i].ReportItem.Exchange, items[i].ReportItem.Asset, items[i].ReportItem.Currency),
+			LinePlots: plots,
+		})
+	}
+
+	return response, nil
+}
+
+// createPNLCharts shows a running history of all realised and unrealised PNL values
+// over time
+func createPNLCharts(items map[string]map[asset.Item]map[currency.Pair]*statistics.CurrencyPairStatistic) (*Chart, error) {
+	if items == nil {
+		return nil, fmt.Errorf("%w missing currency pair statistics", common.ErrNilArguments)
+	}
+	response := &Chart{}
+	for exch, assetMap := range items {
+		for item, pairMap := range assetMap {
+			for pair, result := range pairMap {
+				id := fmt.Sprintf("%v %v %v",
+					exch,
+					item,
+					pair)
+				uPNLName := fmt.Sprintf("%v Unrealised PNL", id)
+				rPNLName := fmt.Sprintf("%v Realised PNL", id)
+
+				unrealisedPNL := ChartLine{Name: uPNLName}
+				realisedPNL := ChartLine{Name: rPNLName}
+				for i := range result.Events {
+					if result.Events[i].PNL != nil {
+						realisedPNL.LinePlots = append(realisedPNL.LinePlots, LinePlot{
+							Value:     result.Events[i].PNL.GetRealisedPNL().PNL.InexactFloat64(),
+							UnixMilli: result.Events[i].Time.UnixMilli(),
+						})
+						unrealisedPNL.LinePlots = append(unrealisedPNL.LinePlots, LinePlot{
+							Value:     result.Events[i].PNL.GetUnrealisedPNL().PNL.InexactFloat64(),
+							UnixMilli: result.Events[i].Time.UnixMilli(),
+						})
+					}
+				}
+				if len(unrealisedPNL.LinePlots) == 0 || len(realisedPNL.LinePlots) == 0 {
+					continue
+				}
+				response.Data = append(response.Data, unrealisedPNL, realisedPNL)
+			}
+		}
+	}
+	return response, nil
+}
+
+// createFuturesSpotDiffChart highlights the difference in futures and spot prices
+// over time
+func createFuturesSpotDiffChart(items map[string]map[asset.Item]map[currency.Pair]*statistics.CurrencyPairStatistic) (*Chart, error) {
+	if items == nil {
+		return nil, fmt.Errorf("%w missing currency pair statistics", common.ErrNilArguments)
+	}
+	var currs []linkCurrencyDiff
+	response := &Chart{}
+	for _, assetMap := range items {
+		for item, pairMap := range assetMap {
+			if !item.IsFutures() {
+				continue
+			}
+			for pair, result := range pairMap {
+				currs = append(currs, linkCurrencyDiff{
+					FuturesPair:   pair,
+					SpotPair:      result.UnderlyingPair,
+					FuturesEvents: result.Events,
+				})
+			}
+		}
+	}
+	for _, assetMap := range items {
+		for item, pairMap := range assetMap {
+			if item.IsFutures() {
+				continue
+			}
+			for pair, result := range pairMap {
+				for i := range currs {
+					if pair.Equal(currs[i].SpotPair) {
+						currs[i].SpotEvents = result.Events
+					}
+				}
+			}
+		}
+	}
```

Comment:
merge for if else handling?

Reply:
Its not that simple, but I'll look to rewrite the whole block

Reply:
😬